libbeat/features/features.go

```diff
@@ -88,30 +92,34 @@ func UpdateFromConfig(c *conf.C) error {
 }
 
 func (f *fflags) SetFQDNEnabled(newValue bool) {
-	f.mu.Lock()
-	defer f.mu.Unlock()
+	f.callbackMut.Lock()
+	defer f.callbackMut.Unlock()
 
+	f.stateMut.Lock()
 	oldValue := f.fqdnEnabled
 	f.fqdnEnabled = newValue
+	f.stateMut.Unlock()
+
 	for _, cb := range f.fqdnCallbacks {
 		cb(newValue, oldValue)
 	}
+
 }
 
 // FQDN reports if FQDN should be used instead of hostname for host.name.
 // If it hasn't been set by UpdateFromConfig or UpdateFromProto, it returns false.
 func FQDN() bool {
-	flags.mu.RLock()
-	defer flags.mu.RUnlock()
+	flags.stateMut.RLock()
+	defer flags.stateMut.RUnlock()
 	return flags.fqdnEnabled
 }
 
 // AddFQDNOnChangeCallback takes a callback function that will be called with the new and old values
 // of `flags.fqdnEnabled` whenever it changes. It also takes a string ID - this is useful
 // in calling `RemoveFQDNOnChangeCallback` to de-register the callback.
 func AddFQDNOnChangeCallback(cb boolValueOnChangeCallback, id string) error {
```

Comment:
Can't there ever be more than one `add_host_metadata` processor globally, e.g. for different pipelines? It looks like their callbacks would overwrite each other in that case, am I reading that right?

Reply:
Hmm, yah, I think you're right. Probably worth adding a check, at the very least.